### ISSUE TITLE
Document which metric name (if any) is retained by comparison operators in different label matching scenarios

### ISSUE THREAD

**Issue by charleskorn** [prometheus/prometheus#15471]
### What did you do?

I ran a query like `left_side > on(env) group_right(dc) right_side` or `left_side > ignoring(pod, region, dc) group_right(dc) right_side` with data like the following (in the PromQL test scripting format):

```
  left_side{env="test", dc="1"}                2 _ _ 5 _
  left_side{env="test", dc="2"}                _ 3 _ _ 6
  right_side{env="test", pod="a", region="au"} 1 2 3 9 10
```

### What did you expect to see?

I expected all returned series to have name `left_side`.

### What did you see instead? Under which circumstances?

All returned series had name `right_side`.

### System information

n/a

### Prometheus version

```text
Prometheus 3
```

### Prometheus configuration file

```yaml
n/a
```

### Alertmanager version

```text
n/a
```

### Alertmanager configuration file

```yaml
n/a
```

### Logs

```text
n/a
```

**Comment by charleskorn** [comment]
This may be related to https://github.com/prometheus/prometheus/issues/5326.

This is a failing test in the PromQL test scripting language that illustrates the issue I'm seeing:

```
load 6m
  left_side{env="test", dc="1"}                2 _ _ 5 _
  left_side{env="test", dc="2"}                _ 3 _ _ 6
  right_side{env="test", pod="a", region="au"} 1 2 3 9 10

# This case currently returns series with name right_side, not left_side.
eval range from 0 to 24m step 6m left_side > on(env) group_right(dc) right_side
  left_side{env="test", pod="a", region="au", dc="1"} 2 _ _ _ _
  left_side{env="test", pod="a", region="au", dc="2"} _ 3 _ _ _

# This evaluates as expected.
eval range from 0 to 24m step 6m left_side > bool on(env) group_right(dc) right_side
  {env="test", pod="a", region="au", dc="1"} 1 _ _ 0 _
  {env="test", pod="a", region="au", dc="2"} _ 1 _ _ 0

# This case currently returns series with name right_side, not left_side.
eval range from 0 to 24m step 6m left_side > ignoring(pod, region, dc) group_right(dc) right_side
  left_side{env="test", pod="a", region="au", dc="1"} 2 _ _ _ _
  left_side{env="test", pod="a", region="au", dc="2"} _ 3 _ _ _
  
# This evaluates as expected.
eval range from 0 to 24m step 6m left_side > bool ignoring(pod, region, dc) group_right(dc) right_side
  {env="test", pod="a", region="au", dc="1"} 1 _ _ 0 _
  {env="test", pod="a", region="au", dc="2"} _ 1 _ _ 0
```

**Comment by beorn7** [comment]
That seems to be an iteration on #5326 . Back then, it was concluded that "when including an on() clause, only labels mentioned in that clause are retained." But with `group_right`, we get the name from the right side maintained after all.

Maybe that's desired and required because the RHS might have multiple different names. If we took the name from the LHS, we would get a collision.

But if it _is_ actually the desired behavior, we have to document it. Of course(?!?), we never documented the behavior deemed as "correct" in #5326.

We need more qualified PromQL experts to make a call here. Paging @juliusv and @roidelapluie .

**Comment by charleskorn** [comment]
I think there are two differences here compared to #5326:

* in this case, the issue applies to both `on` and `ignoring`, whereas #5326 is only about `on`
* in this case, the metric name from the right-hand side is being returned, whereas the behaviour in #5236 is to not include the metric name from either side

**Comment by juliusv** [comment]
> I think there are two differences here compared to #5326:
> 
> * in this case, the issue applies to both `on` and `ignoring`, whereas [Vector/vector filter comparisons return match group labels rather than LHS labels #5326](https://github.com/prometheus/prometheus/issues/5326) is only about `on`

I think that was just an oversight by me back then, and the same issue would have applied to `ignoring`.

> * in this case, the metric name from the right-hand side is being returned, whereas the behaviour in [Vector/vector filter comparisons return match group labels rather than LHS labels #5326](https://github.com/prometheus/prometheus/issues/5326) is to not include the metric name from either side

When there is many-to-one (`group_left`) or one-to-many (`group_right`) matching involved, you will always have the extra dimensionality labels from the "many" side in the output, in addition to the matching labels (since you would otherwise have multiple colliding output time series with the same label set).

I guess the question is only if the `__name__` label should be treated specially here and removed. I think probably not, because sometimes it may be the only distinguishing feature. Let's say you do:

```
a > on(label1) group_right {__name__=~"b|c"}
```

With data like:

```
a{label1="foo"}
a{label1="bar"}

b{label1="foo"}
b{label1="bar"}

c{label1="foo"}
c{label1="bar"}
```

That expression would then currently return:

```
b{label1="foo"}
b{label1="bar"}
c{label1="foo"}
c{label1="bar"}
```

But if the metric name got auto-removed, you would get a collision (and thus an error):

```
{label1="foo"}
{label1="bar"}
{label1="foo"}
{label1="bar"}
```

We also don't have any special modifier in `on()`/`ignoring()`/`group_x()` that would let you control this behavior, so I guess it should probably keep the metric name, in absence of such a modifier. Of course one can still always remove the metric name manually before (or after) the filtering, if it's not desired.

**Comment by charleskorn** [comment]
>I think that was just an oversight by me back then, and the same issue would have applied to `ignoring`.

This test currently passes, which suggests to me that the metric name is only dropped when `on` is used, but not `ignoring`:

```
load 5m
    series_a{env="test", foo="foo"}    0 1 2 3
    series_a{env="prod", foo="bar"}    4 5 6 7
    series_b{env="test"}               0 0 2 2
    series_b{env="prod"}               4 4 6 6

eval range from 0 to 15m step 5m series_a == on (env) series_b
    {env="test"} 0 _ 2 _
    {env="prod"} 4 _ 6 _

eval range from 0 to 15m step 5m series_a == ignoring (foo) series_b
    series_a{env="test"} 0 _ 2 _
    series_a{env="prod"} 4 _ 6 _
```



> > * in this case, the metric name from the right-hand side is being returned, whereas the behaviour in [Vector/vector filter comparisons return match group labels rather than LHS labels #5326](https://github.com/prometheus/prometheus/issues/5326) is to not include the metric name from either side
> 
> When there is many-to-one (`group_left`) or one-to-many (`group_right`) matching involved, you will always have the extra dimensionality labels from the "many" side in the output, in addition to the matching labels (since you would otherwise have multiple colliding output time series with the same label set).
> 
> I guess the question is only if the `__name__` label should be treated specially here and removed. I think probably not, because sometimes it may be the only distinguishing feature. 

Sorry, perhaps I wasn't clear:

* when `group_left` is used, the metric name from the left series is currently returned, as I would expect.
* when `group_right` is used, the metric name from the right series is currently returned, rather than the metric name from the left series, which is what I'd expect. I don't think we should remove the metric name in this case. 
* when neither `group_left` or `group_right` is used, then I see the behaviour from https://github.com/prometheus/prometheus/issues/5326 where the metric name is dropped when `on` is used, and retained when `ignoring` is used.

With the fix from https://github.com/prometheus/prometheus/pull/15473, `group_right` will return the metric name from the left series, rather than the right series. 

This would cause a conflict in the example you provided, because the output series would be this:

```
a{label1="foo"}
a{label1="bar"}
a{label1="foo"}
a{label1="bar"}
```

**Comment by juliusv** [comment]
> > I think that was just an oversight by me back then, and the same issue would have applied to `ignoring`.
> 
> This test currently passes, which suggests to me that the metric name is only dropped when `on` is used, but not `ignoring`:

Yes, but this is not because we are explicitly treating the metric name in a special and different way in `on` vs. `ignoring`. The `on` and `ignoring` modifiers are just alternate ways of specifying a list of labels to match on. The labels that make it into the result then are:

* All matching labels (which are defined either explicitly via `on()` or implicitly via `ignoring()`)
* Any extra dimensionality labels from the "many" side

Since the metric name is not special here, but just another label with the label name `__name__`, it just ends up being one of the "many" side labels that are included in the result when using either `group_left` or `group_right`.

> Sorry, perhaps I wasn't clear:
> 
> * when `group_left` is used, the metric name from the left series is currently returned, as I would expect.
> * when `group_right` is used, the metric name from the right series is currently returned, rather than the metric name from the left series, which is what I'd expect. I don't think we should remove the metric name in this case.
> * when neither `group_left` or `group_right` is used, then I see the behaviour from [Vector/vector filter comparisons return match group labels rather than LHS labels #5326](https://github.com/prometheus/prometheus/issues/5326) where the metric name is dropped when `on` is used, and retained when `ignoring` is used.

Yep, see my explanation above. I don't know if we can do anything better here without rethinking things completely (and breaking them). Not only would collisions result as in my example use case, but we would also be changing long-established behavior that people may rely on.

What I'd do in practice maybe to make such query outputs less confusing is to remove the metric name on the "many" side completely (that is, if it is not required for series uniqueness).

**Comment by charleskorn** [comment]
What you're describing makes sense to me from the perspective of maintaining the existing behaviour, but it differs from what the documentation says should happen.

Perhaps a good next step would be to update the documentation to match what you've described?

**Comment by beorn7** [comment]
Wow, that went really deep… :exploding_head: 

From my understanding, I would say we should keep the existing behavior, not only to avoid breaking the expectation that has grown over the years, but also because the current behavior is one possibility of the many that can be considered "reasonable".

And yes, we should document it, but in a way that doesn't totally confuse the readers that aren't really interested in these intricacies.

**Comment by juliusv** [comment]
> What you're describing makes sense to me from the perspective of maintaining the existing behaviour and not breaking things that assume this behaviour, but it differs from what the documentation says should happen.

Looking at https://prometheus.io/docs/prometheus/latest/querying/operators/#operators, things look mostly correct in terms of the descriptions for vector-to-vector filter operators, except that https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches should probably explicitly mention that the output label set will not only contain the matching labels (as mentioned earlier in https://prometheus.io/docs/prometheus/latest/querying/operators/#comparison-binary-operators), but also the extra labels from the "many" side. Or maybe there was also something else you spotted that I missed?

**Comment by charleskorn** [comment]
I think the "Between two instant vectors" paragraph under https://prometheus.io/docs/prometheus/latest/querying/operators/#comparison-binary-operators needs some expansion. The last sentence is currently:

> The metric name is dropped if the `bool` modifier is provided.

This implies that the metric name isn't dropped if `bool` isn't provided, which doesn't match what we've described above. Perhaps something along these lines would clarify the behaviour?

> The metric name is dropped if the `bool` modifier is provided.
> The metric name from the left side is retained if the `bool` modifier is not provided, with some exceptions:
> 
> * If `on` is used with one-to-one matching, the metric name is dropped.
> * If `group_right` is used, the metric name from the right side (not the left) is retained.

**Comment by beorn7** [comment]
> If on is used with one-to-one matching, the metric name is dropped.

… unless you say `on (__name__)`. I think the explanation should be different: If you use `on` or `ignoring`, only the labels the match in performed on are retained. This rule includes the `__name__` label.

> If group_right is used, the metric name from the right side (not the left) is retained.

While this describes the outcome correctly, I think the real explanation is that the labels used for the grouping _have_ to be returned to avoid collisions, and with `group_right`, the name of the RHS is used for grouping.

In other words, maybe we can find a wording where the exceptions above are actually not exceptions but covered by the rules.

**Comment by charleskorn** [comment]
> I think the explanation should be different: If you use `on` or `ignoring`, only the labels the match in performed on are retained. This rule includes the `__name__` label.

I think this isn't quite right either: if you use `ignoring`, then the metric name isn't used for matching, but it will appear in the result.

**Comment by charleskorn** [comment]
I've put together https://github.com/prometheus/prometheus/pull/16716 for discussion.

**Comment by bboreham** [comment]
Hello from the bug-scrub!  Since #16716 is merged we will close this.  Please reopen and clarify what is left to do here if necessary.
